hw/ip/otbn/dv/rig/rig/gens/bad_load_store.py

```diff
@@ -0,0 +1,211 @@
+                self.bnsid.operands[1].op_type.reg_type == 'gpr' and
+                not self.bnsid.operands[1].op_type.is_dest() and
+                isinstance(self.bnsid.operands[2].op_type, ImmOperandType) and
+                self.bnsid.operands[2].op_type.signed and
+                isinstance(self.bnsid.operands[3].op_type, OptionOperandType) and
+                isinstance(self.bnsid.operands[4].op_type, OptionOperandType)):
+            raise RuntimeError('BN.LID instruction from instructions file is '
+                               'not the shape expected by the BadLoadStore generator.')
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+        # Return None if this is the last instruction in the current gap
+        # because we need to either jump or do an ECALL to avoid getting stuck.
+        if program.get_insn_space_at(model.pc) <= 1:
+            return None
+
+        op_val_grd = model.pick_operand_value(self.grd_op_type)
+        assert op_val_grd is not None
+
+        offset_rng = self.imm_op_type.get_op_val_range(model.pc)
+        assert offset_rng is not None
+
+        bn_offset_rng = self.bn_imm_op_type.get_op_val_range(model.pc)
+        assert bn_offset_rng is not None
+
+        # Pick a really negative or really positive known value for hitting OOB
+        base = []
+        wdr_gprs = []
+
+        known_regs = model.regs_with_known_vals('gpr')
+        for reg_idx, reg_val in known_regs:
+            if 4294965246 > reg_val > 2147485695:
+                base.append((reg_idx, reg_val))
+            if 2147481600 > reg_val > 6143:
+                base.append((reg_idx, reg_val))
+            if (6140 >= reg_val >= 2050):  # If reg_val reach very close to the top
+                base.append((reg_idx, reg_val))
+            if reg_val <= 31:  # Pick a register which could be suitable for bn.xid
+                wdr_gprs.append(reg_val)
+        if not base:
+            return None
+
+        # Pick a random register among known registers that is constrained above
+        idx, value = random.choice(base)
+        if value <= 6140:
+            imm_val = 4096 - value
+        else:
+            imm_val = random.choice(offset_rng)
+
+        addr = imm_val + value
+        assert addr is not None
+
+        op_val_grs1 = idx
+        if op_val_grs1 is None:
+            return None
+
+        idx_grs2, value_grs2 = random.choice(known_regs)
+        op_val_grs2 = idx_grs2
+        if op_val_grs2 is None:
+            return None
+
+        offset_val = self.imm_op_type.op_val_to_enc_val(imm_val, model.pc)
+        assert offset_val is not None
+
+        bn_imm_val = random.choice(bn_offset_rng)
+        bn_offset_val = self.bn_imm_op_type.op_val_to_enc_val(bn_imm_val, model.pc)
+        assert bn_offset_val is not None
+
+        wdr_gpr_idx = random.choice(wdr_gprs)
+        if wdr_gpr_idx is None:
+            return None
```

Comment:
I think that `wdr_gprs` should never be empty, because it always contains `x0` (so the call to `random.choice` shouldn't explode). Maybe worth adding a comment to explain why? Also, the result will never be `None`.